### PR TITLE
make path_is_valid() stricter, tweak path_extract_filename()

### PR DESCRIPTION
**Pull request by poettering**

### CODE REVIEW

File: src/basic/path-util.c [hunk 1]
```diff
@@ -851,19 +853,21 @@ int path_extract_filename(const char *p, char **ret) {
...
 
-        for (q = c; *q != 0; q++)
+        for (q = c; *q != 0; q++) /* Find trailing slashes */
                 if (*q != '/')
                         e = q + 1;
 
-        if (!e) /* no valid character? */
-                return -EINVAL;
+        if (!e) /* not a single non-slash character, then there is no filename to extract, return recognizable error */
+                return -EADDRNOTAVAIL;
```
Comment (yuwata) [inline src/basic/path-util.c]: How about the following for readability?
```
if (streq(c, "/"))
    return -EADDRNOTAVAIL;

e = strchrnul(c, '/');
```

File: src/test/test-path-util.c [hunk 2]
```diff
@@ -618,16 +617,49 @@ static void test_filename_is_valid(void) {
         assert_se(!filename_is_valid("bar/foo/"));
         assert_se(!filename_is_valid("bar//"));
 
-        for (i=0; i<FILENAME_MAX+1; i++)
+        for (size_t i = 0; i < sizeof(foo)-1; i++)
                 foo[i] = 'a';
-        foo[FILENAME_MAX+1] = '\0';
+        foo[sizeof(foo)-1] = '\0';
```
Comment (keszybz) [inline src/test/test-path-util.c]: `char_array_0` ?

File: src/test/test-path-util.c [hunk 3]
```diff
@@ -618,16 +617,49 @@ static void test_filename_is_valid(void) {
...
+        assert_se(path_is_valid("/////.///.////...///..//."));
+        assert_se(path_is_valid("."));
+        assert_se(path_is_valid(".."));
+        assert_se(path_is_valid("bar/foo"));
+        assert_se(path_is_valid("bar/foo/"));
+        assert_se(path_is_valid("bar//"));
+
+        for (size_t i = 0; i < sizeof(foo)-1; i++)
+                foo[i] = 'a';
+        foo[sizeof(foo)-1] = '\0';
```
Comment (keszybz) [inline src/test/test-path-util.c]: Ditto.

**Review by keszybz — COMMENTED** [review]
> This changes API, but none of the existing callers care about the return
value, hence the change should be fine.

Is this in preparation for something?

**Comment by poettering** [comment]
> Is this in preparation for something?

(yeah, i have been toying around with a minimal updating tool the past days, that reuses importd downloaders to write to partitions.)

**Comment by poettering** [comment]
Force pushed a new version. I changed both functions a bit around, hence, please review again.

File: src/basic/path-util.c (hunk 1 above)
Comment (poettering) [reply src/basic/path-util.c]: made a similar change. ptal

File: src/basic/path-util.c [hunk 4]
```diff
@@ -823,6 +823,8 @@ const char *last_path_component(const char *path) {
          *    Also, the empty string is mapped to itself.
          *
          * This is different than basename(), which returns "" when a trailing slash is present.
+         *
+         * This always succeeds (except if you pass NULL in which case it returns NULL, too)
```
Comment (yuwata) [inline src/basic/path-util.c]: Not a matter, but period is missing.

**Review by yuwata — COMMENTED** [review]
I think it is better that `path_extract_filename()` also supports `/aa/bb/. -> bb`. What do you think about applying `path_simplify(, true)` at first?

Moreover, `/aa/bb/.. -> aa`??

**Comment by poettering** [comment]
> I think it is better that `path_extract_filename()` also supports `/aa/bb/. -> bb`. What do you think about applying `path_simplify(, true)` at first?

dunno, i think it woud be quite untypical to specify, no? the current logic 

> 
> Moreover, `/aa/bb/.. -> /aa`??

This actually isn't the same. If bb is a symlink to `/cc` then `/aa/bb/..` actualy resolves to `/`, not `/aa`. path_simplify() won't remove `..` away either because of that. unless we look at the disk we can't determine what `..` will do

**Comment by yuwata** [comment]
> in general i'd probably refuse if in doubt, and then see what happens.

That makes sense.

**Comment by poettering** [comment]
added the suggested full stop. thanks. 

three CI jobs failed for this, all about timeouts, two about fuzzing timeouts. I wonder if these are flukes, or this PR is indeed buggy. force pushed rebased version to trigger the CIs again, so that we can see.

**Comment by poettering** [comment]
The CI failures puzzle me. I split this in two PRs now, #18437 and #18438. Let's see which one is at fault

**Comment by keszybz** [comment]
Let's close this in favour of #18437 and/or #18438.

**Pull request closed without merging**